odl/tomo/operators/spect_trafo.py

```diff
@@ -0,0 +1,226 @@
+
+import numpy as np
+
+# Internal
+import odl
+from odl.operator.operator import Operator
+
+
+_SUPPORTED_IMPL = ('niftyrec')
+__all__ = ('SpectProject', 'SpectBackproject')
+
+
+class SpectProject(Operator):
+    """Spect projector using backend NiftyRec"""
+    def __init__(self, dom, geometry, attenuation, psf, use_gpu=False,
+                 impl='niftyrec', **kwargs):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        dom : `DiscreteLp`
+            Discretized space, the domain of the forward projector
+        geometry : `Geometry`
+            The geometry of the transform. An ODL geometry instance
+            that contains information about the operator domain
+        attenuation : `DiscreteLpVector`
+            Linear attenuation map for SPECT.
+        psf : `DiscreteLpVector`
+            Point Spread Functions for the correction of the
+            collimator-detector response
+        use_gpu : `bool`, optional
+            'False' if not using GPU
+            `True` if using GPU provided that `impl` supports it
+        impl='niftyrec', optional
+            Implementation back-end for the transform. Supported back-ends:
+            NiftyRec Tomography Toolbox through NiftyPy a Python wrapper
+            for NiftyRec
+        """
+        impl, impl_in = str(impl).lower(), impl
+        if impl not in _SUPPORTED_IMPL:
+            raise ValueError('implementation {!r} not supported.'
+                             ''.format(impl_in))
+
+        self.geometry = geometry
+        det_nx = geometry.det_grid.shape[0]
+        det_ny = geometry.det_grid.shape[1]
+        startAngle = float(geometry.motion_grid.points()[0])
+        stopAngle = float(geometry.motion_grid.points()[-1]) * (180 / np.pi)
+        N_proj = len(geometry.motion_grid.points())
+
+        self._cameras = np.float32(np.linspace(startAngle, stopAngle, N_proj,
+                                   dtype=np.float32).reshape((N_proj, 1)))
+        self._use_gpu = use_gpu
+        self._attenuation = attenuation
+        self._psf = psf
+
+        lowleft_x = geometry.detector.params.corners()[0][0]
+        lowleft_y = geometry.detector.params.corners()[0][1]
+        upright_x = geometry.detector.params.corners()[-1][0]
+        upright_y = geometry.detector.params.corners()[-1][1]
+        ran = odl.uniform_discr([lowleft_x, lowleft_y, startAngle],
+                                [upright_x, upright_y, stopAngle],
+                                [det_nx, det_ny, N_proj])
+
+        super().__init__(dom, ran, linear=True)
+
+    def _call(self, x):
+        """Compute the discrete SPECT ptojection
+
+        Parameters
+        ----------
+        x : `DiscreteLpVector`
+           Element in the domain of the operator to be forward projected
+
+        Returns
+        -------
+        out : `DiscreteLpVector`
+            Returns an element in the projection space
+        """
+        if self._attenuation is None:
```

Comment:
Because of the following `else` statement and afterwards this "no-longer ODL space element" is passed to the `SpectBackproject` which requires `attenuation` to be a `element in :class: ~odl.discr.lp_discr.DiscreteLp`